105-strimzi-drain-cleaner-namespace-filtering-proposal.md

```diff
@@ -0,0 +1,256 @@
+
+### Real-World Impact
+
+This issue affects enterprise environments where:
+- Kafka is deployed in specific namespaces (e.g., `kafka-prod`, `kafka-dev`, `messaging`)
+- Security teams enforce least-privilege RBAC policies
+- Compliance requirements mandate namespace-level access controls
+- One drain cleaner instance must serve the entire cluster while respecting security boundaries
+
+### Potential Negative Impacts
+
+**Risk of Multiple Drain Cleaner Deployments**: With namespace filtering, users might be tempted to deploy multiple drain cleaner instances per cluster, which could lead to:
+- Conflicting ValidatingWebhookConfigurations
+- Duplicate webhook processing and potential race conditions
+- Increased operational complexity and resource usage
+- Certificate management challenges
+
+**Mitigation**: Clear documentation must emphasize that:
+- Only ONE drain cleaner instance should be deployed per cluster
+- Namespace filtering is for security compliance, not for multi-tenancy
+- Multiple instances will cause operational issues and are not supported
+
+## Goals
+
+- **Primary Goal**: Enable namespace-level filtering to allow drain cleaner deployment in restrictive RBAC environments
+- **Security Goal**: Reduce actual pod API calls to only configured namespaces while maintaining webhook functionality
+- **Operational Goal**: Eliminate log noise from permission errors in unmanaged namespaces
+- **Compatibility Goal**: Maintain full backward compatibility with existing deployments
+
+## Non-Goals
+
+- Modifying cluster-wide webhook registration (webhooks must still intercept all eviction requests)
+- Changing RBAC configurations or permissions automatically
+- Supporting multiple drain cleaner instances per cluster
+- Automatic migration or configuration discovery
+
+## Proposal
+
+### Overview
+
+Introduce a new environment variable `STRIMZI_DRAIN_NAMESPACES` that accepts a comma-separated list of namespace names. When configured, the drain cleaner will:
+
+1. Continue receiving all eviction webhook requests (required for webhook functionality)
+2. Filter requests at the application level before attempting pod access
+3. For unwatched namespaces: immediately allow eviction without pod API calls (logged at DEBUG level)
+4. For watched namespaces: process normally with full drain cleaner logic
+
+### Configuration
+
+**Environment Variable**: `STRIMZI_DRAIN_NAMESPACES`
+- **Format**: Comma-separated namespace list (e.g., `"kafka-prod,kafka-dev,messaging"`) or `"*"` for all namespaces
+- **Default**: `"*"` (process all namespaces - current behavior)
+- **Behavior**:
+  - `"*"` or unset: Process all namespaces (backward compatible)
+  - Specific namespace list: Only process eviction requests from specified namespaces
+
+### Application-Level Filtering
+
+The filtering occurs in the `ValidatingWebhook.webhook()` method:
+
+```java
+private boolean isNamespaceWatched(String namespace) {
```

Comment:
For the implementation, will the Namespaces that Drain Cleaner should operate in be parsed somewhere else?
I mean, it would be better to have it processed somewhere else, so you don't do the parsing every time in the `isNamespaceWatched` method.
But if this snippet is just for example purposes, then it's fine :)

Reply:
Thank you very much for the review and remarks! 
I've just updated proposed realisation.